src/audio/sink_api_helper.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/sink_api.h>
+#include <sof/audio/sink_api_implementation.h>
+
+void sink_init(struct sink __sparse_cache *sink, const struct sink_ops *ops,
+	       struct sof_audio_stream_params *audio_stream_params)
+{
+	sink->ops = ops;
+	sink->audio_stream_params = audio_stream_params;
+}
+
+size_t sink_get_free_size(struct sink __sparse_cache *sink)
+{
+	return sink->ops->get_free_size(sink);
+}
+
+int sink_get_buffer(struct sink __sparse_cache *sink, size_t req_size,
+		    void __sparse_cache **data_ptr, void __sparse_cache **buffer_start,
+		    size_t *buffer_size)
+{
+	int ret;
+
+	if (sink->locked_write_frag_size)
+		return -EBUSY;
+
+	ret = sink->ops->get_buffer(sink, req_size, data_ptr,
+						buffer_start, buffer_size);
+
+	if (!ret)
+		sink->locked_write_frag_size = req_size;
+	return ret;
+}
+
+int sink_commit_buffer(struct sink __sparse_cache *sink, size_t commit_size)
+{
+	int ret;
+
+	/* check if there was a buffer obtained for writing by sink_get_buffer */
+	if (!sink->locked_write_frag_size)
+		return -ENODATA;
+
+	/* limit size of data to be committed to previously obtained size */
+	if (commit_size > sink->locked_write_frag_size)
+		commit_size = sink->locked_write_frag_size;
+
+	ret = sink->ops->commit_buffer(sink, commit_size);
+
+	if (!ret)
+		sink->locked_write_frag_size = 0;
+
+	sink->num_of_bytes_processed += commit_size;
+	return ret;
+}
+
+size_t sink_get_num_of_processed_bytes(struct sink __sparse_cache *sink)
```

Comment:
i feel like all of these should be annotate with `inline`

Reply:
this seems to be an exported function, so it cannot be `inline`

Reply:
It may be, it won't hurt. Theoretically is possible to inline exported function, however, I don't think clang compiler/linker is smart enough to do it.

Reply:
Yeah, that's a LTO feature, and nothing has wired it up for Zephyr builds yet (and in any case the Xtensa clang is based on an LLVM too old to support it anyway).

There's lots of aesthetic argument over inline in C.  FWIW, conventions within the Zephyr tree itself (whether SOF wants to honor that is yet another argument, heh) have largely fallen out as:

* Use "static inline" liberally as a trick to allow function bodies to appear in header files and achieve cheap "LTO-lite" advantages for simple APIs (at the cost of the occasional horrifying knot of header dependency hell, to be fair).

* Generally don't try to use inline for performance choices within a single C file.  Modern compilers are probably smarter about this than you are.  If you are tempted, at least look at the generated code first to be sure you know what you want and what you are getting.

* If you are at the point where you're looking at generated code and hyperoptimizing, prefer the `ALWAYS_INLINE` macro instead, which is a firmer promise that Zephyr provides for all toolchains[1].

[1] There are a tiny handful of spots where we rely on this for correctness reasons, e.g. a macro-like "C function" in the arch layer that needs to inspect the stack pointer and will behave differently if it's in a different frame.  There are situations in Linux where this is needed for memory barrier control too, IIRC.

Reply:
Oh, and to jump in on the export bit: there's no particular limitation there.  Obviously a function **called** from an external translation unit can only be a true function, but within the file the compiler is free to inline if it wants to, subject to whatever the heuristics tell it is a good idea (being a visible symbol, it's going to be required to be present in the output assembly anyway, so -Os might say "no inlining" for that reason, etc...)